### PR TITLE
[Peripherals] Fix segfault by adding some synchronization to peripheral add-ons

### PR DESCRIPTION
**Pull request by garbear**
This fixes a segfault at exit when peripheral add-ons misbehave. In particular, if the add-on hangs in another thread, then it can be destroyed while still in use. This was causing a segfault on exit.

Another segfault (or possibly the same one) has been fixed by moving g_peripherals to ServiceManager. Before, it tried to release resources when the static object was deconstructed after the app shut down.

## Types of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

**Comment by garbear** [comment]
I've pushed a new commit and updated the PR description. It turns out, there were only 2 segfaults at work here, one from misbehaving peripheral add-ons, and one due to g_peripherals being a static variable and trying to release resources that had already disappeared.

**Comment by MartijnKaijser** [comment]
Does this also affect v17? 
Jenkins build this please

**Comment by garbear** [comment]
Yes, peripheral.joystick is still hanging on the Steam Link, so without this Kodi crashes on exit 100% of the time.

**Comment by garbear** [comment]
This almost concludes my Krypton backports. The only other bugs I care to fix are the iMON incompatibility and better Shield controller support. If we're gearing up for 17.1 release I'm fine waiting for 17.2 or 18.0.

**Comment by garbear** [comment]
Hold off on merging. The shutdown stuff still isn't happening in the right order.

**Comment by garbear** [comment]
lol, peripherals shutdown was 1 line off. Also pushed a commit for a third segfault that arose from starting/stopping things in the proper order.

jenkins build this please

**Comment by garbear** [comment]
jenkins build this please

File: xbmc/games/ports/PortMapper.cpp [hunk 1]
```diff
@@ -23,13 +23,18 @@
...
 using namespace KODI;
 using namespace GAME;
 using namespace JOYSTICK;
 using namespace PERIPHERALS;
 
-CPortMapper::CPortMapper()
+CPortMapper::CPortMapper(PERIPHERALS::CPeripherals* peripheralManager) :
+  m_peripheralManager(peripheralManager)
 {
+  assert(m_peripheralManager != nullptr);
```
Comment (thebrid) [inline xbmc/games/ports/PortMapper.cpp]: If null doesn't make sense here, should we pass by reference rather than pointer?

Comment (garbear) [inline xbmc/games/ports/PortMapper.cpp]: Yes, passing by reference would be better. I'll update the PR.

**Comment by garbear** [comment]
updated with @thebrid's comment. jenkins build this please

**Comment by garbear** [comment]
Build errors look unrelated, but both OSX and linux64 failed, so again we try. Jenkins build this please.

**Comment by eigendude** [comment]
third time's the charm. jenkins build and merge

**Comment by garbear** [comment]
let's try again. jenkins build and merge

**Comment by wsnipex** [comment]
jenkins build this please

**Comment by garbear** [comment]
There's an error with the test cases. Moving around the order of services on startup typically breaks things. I'll address this when I get a chance. Until then, this problem only manifests with broken peripheral add-ons, which is only peripheral.joystick on the Steam Link (and I work around it). So not critical.

**Comment by garbear** [comment]
test cases fixed! jenkins build this please

**Pull request merged**